packages/lisk-transactions/src/utils/sign_and_verify.ts

```diff
@@ -84,51 +85,74 @@ export const verifyMultisignatures = (
 			errors: [new TransactionError('Sender does not have valid multimin', id)],
 		};
 	}
-	if (!signatures || (signatures && signatures.length < minimumValidations)) {
+	if (!Array.isArray(signatures)) {
 		return {
 			verified: false,
 			errors: [new TransactionError('Missing signatures', id, '.signatures')],
 		};
 	}
+	if (signatures.length > MAX_MULTISIG_SIGNATURES) {
+		return {
+			verified: false,
+			errors: [
+				new TransactionError(
+					'Exceeds maximum of 15 signatures.',
+					id,
+					'.signatures',
+				),
+			],
+		};
+	}
 
-	const verifiedSignatures: string[] = [];
-	const checkedPublicKeys: string[] = [];
+	const checkedPublicKeys = new Map();
+	const verifiedSignatures = new Map();
 
 	memberPublicKeys.forEach(publicKey => {
-		if (checkedPublicKeys.includes(publicKey)) {
+		if (checkedPublicKeys.has(publicKey)) {
 			return;
 		}
 
 		signatures.forEach((signature: string) => {
-			if (verifiedSignatures.includes(signature)) {
+			if (verifiedSignatures.has(signature)) {
 				return;
 			}
 			const { verified: signatureVerified } = verifySignature(
 				publicKey,
 				signature,
 				transactionBytes,
+				id,
 			);
 			if (signatureVerified) {
-				checkedPublicKeys.push(publicKey);
-				verifiedSignatures.push(signature);
+				checkedPublicKeys.set(publicKey, true);
+				verifiedSignatures.set(signature, true);
 			}
 		});
 	});
 
-	const unverifiedSignatures = signatures.filter(
-		signature =>
-			!verifiedSignatures.find(
-				verifiedSignature => signature === verifiedSignature,
-			),
+	const invalidSignatures = signatures.filter(
+		signature => !verifiedSignatures.has(signature),
 	);
 
+	// Transaction is waiting for more signatures
+	if (
+		signatures.length < minimumValidations &&
+		invalidSignatures.length === 0
+	) {
+		return {
+			verified: false,
+			errors: [
+				new TransactionPendingError(`Missing signatures`, id, '.signatures'),
+			],
+		};
+	}
+
 	return {
 		verified:
-			verifiedSignatures.length >= minimumValidations &&
-			unverifiedSignatures.length === 0,
+			verifiedSignatures.size >= minimumValidations &&
```

Comment:
Assuming there is no invalidSignature` at this point, # of `verifiedSignature` should equal `signatures here.
so probably checking `signatures.length >= minimumValidations` is sufficient?